File: Algorithm/BusRoutes.cpp
// Solution: BFS
// Time complexity: O(m*n) m: # of buses, n: # of routes
// Space complexity: O(m*n + m)
class Solution {
public:
    int numBusesToDestination(vector<vector<int>>& routes, int S, int T) {
        if (S == T) return 0;
        
        unordered_map<int, vector<int>> m;
        for (int i = 0; i < routes.size(); ++i)
            for (const int stop : routes[i])
                m[stop].push_back(i);
        
        vector<int> visited(routes.size(), 0);
        queue<int> q;
        q.push(S);
        int buses = 0;
        
        while (!q.empty()) {
            int size = q.size();
            ++buses;
            while (size--) {
                int curr = q.front(); q.pop();
                for (const int bus : m[curr]) {
                    if (visited[bus]) continue;
                    visited[bus] = 1;
                    for (int stop : routes[bus]) {
                        if (stop == T) return buses;
                        q.push(stop);
                    }
                }
            }
        }
        return -1;
    }
};

File: Algorithm/BusRoutes.java
// Time complexity: O(m * n) 公交车数量 * 每辆公交车路线stop数量
// Space complexity: O(m * n)
class Solution {
    public int numBusesToDestination(int[][] routes, int S, int T) {
        if (S == T) return 0;
        
        Map<Integer, List<Integer>> m = new HashMap<>(); // stop -> {bus}
        for (int i = 0; i < routes.length; ++i) 
            for (final int stop : routes[i]) 
                m.computeIfAbsent(stop, k -> new ArrayList<Integer>()).add(i);
            
        int[] ride = new int[routes.length];
        Queue<Integer> q = new LinkedList<Integer>(); // 标记当前要扩展的stop Id。
        q.offer(S);
        int buses = 0;
        
        while (!q.isEmpty()) {
            int size = q.size();
            ++buses;
            while (size-- != 0) {
                int curr = q.poll();
                for (final Integer bus : m.get(curr)) {
                    if (ride[bus] == 1) continue;
                    ride[bus] = 1;
                    for (int stop : routes[bus]) {
                        if (stop == T) return buses;
                        q.offer(stop);
                    }
                }
            }
        }
        return -1;
    }
}
